services/offline/src/lib/cacheable-section-state.js

```diff
@@ -0,0 +1,165 @@
+// state in a performant way
+
+/**
+ * Create a store for Cacheable Section state.
+ * Expected to be used in app adapter
+ */
+export function createCacheableSectionStore() {
+    const initialState = { recordingStates: {}, cachedSections: {} }
+    return createStore(initialState)
+}
+
+/**
+ * Provides context for a global state context which will track cached
+ * sections' status and cacheable sections' recording states, which will
+ * determine how that component will render. The provider will be a part of
+ * the OfflineProvider.
+ */
+export function CacheableSectionProvider({ children, store }) {
+    const offlineInterface = useOfflineInterface()
+
+    // On load, get sections and add to store
+    React.useEffect(() => {
+        offlineInterface.getCachedSections().then(sections => {
+            const newSections = sections.reduce(
+                (result, { sectionId, lastUpdated }) => {
+                    return { ...result, [sectionId]: lastUpdated }
+                },
+                {}
+            )
+            store.mutate(state => ({
+                ...state,
+                cachedSections: newSections,
+            }))
+        })
+    }, [store, offlineInterface])
+
+    return <GlobalStateProvider store={store}>{children}</GlobalStateProvider>
+}
+CacheableSectionProvider.propTypes = {
+    children: PropTypes.node,
+    store: PropTypes.shape({ mutate: PropTypes.func }),
+}
+
+/**
+ * Uses an optimized global state to manage 'recording state' values without
+ * unnecessarily rerendering all consuming components
+ *
+ * @param {String} id - ID of the cacheable section to track
+ * @returns {Object} { recordingState: String, setRecordingState: Function, removeRecordingState: Function}
+ */
+export function useRecordingState(id) {
+    const [recordingState] = useGlobalState(state => state.recordingStates[id])
+    const setRecordingState = useGlobalStateMutation(newState => state => ({
+        ...state,
+        recordingStates: { ...state.recordingStates, [id]: newState },
+    }))
+    const removeRecordingState = useGlobalStateMutation(() => state => {
+        const recordingStates = { ...state.recordingStates }
+        delete recordingStates[id]
+        return { ...state, recordingStates }
+    })
+
+    return { recordingState, setRecordingState, removeRecordingState }
+}
+
+/**
+ * Returns a function that syncs cached sections in the global state
+ * with IndexedDB, so that IndexedDB is the single source of truth
+ *
+ * @returns {Function} updateCachedSections
+ */
+function useUpdateCachedSections() {
+    const offlineInterface = useOfflineInterface()
+    const setCachedSections = useGlobalStateMutation(
+        cachedSections => state => ({
+            ...state,
+            cachedSections,
+        })
+    )
+
+    return async function updateCachedSections() {
+        const sections = await offlineInterface.getCachedSections()
+        const newSections = sections.reduce(
```

Comment:
Since this transformation also occurs above, might make sense to extract it as a helper function